### DISCUSSION TITLE
[Pytorch] Create CPP presets for other Pytorch versions

### DISCUSSION THREAD

**Discussion by carlosuc3m** [bytedeco/javacpp-presets#1115] [General]
Hello I am developing an application that is able to load pre-trained Pytorch models and execute inference with them. 
I want to try Javacpp to load Pytorch in Java. I have seen the Javacpp-presets for Pytorch but I assume that they are only available
for Pytorch 1.9.0. Is there any help I can use to try to create the presets for previous versions?
Thank you @saudet for pointing this repo in other forum.
Regards,
Carlos

**Comment by carlosuc3m** [comment]
I just found the documentation:
https://github.com/bytedeco/javacpp-presets/wiki/Create-New-Presets
Just out of curiosity, how does this approach allow loading and unloading of native libraries at runtime?

**Reply by saudet** [reply]
Do you mean like LibTorch? That's possible, yes, instructions for that are here:
https://github.com/bytedeco/javacpp-presets/tree/master/pytorch#documentation

**Reply by saudet** [reply]
If you mean how to unload a library, JavaCPP uses `System.load()` to load them, so whatever works to unload those will work.

**Reply by carlosuc3m** [reply]
Usually unloading a native library is done by Garbage collecting the classloader that loaded the class that performed System.load() the corresponding native library, however it seems that the native libraries are not unloaded.
Here I created an example program (using TEnsorflow presets instead of Pytorch) that loads the native library for a class, tries to unload it and then tries to load another Tensorflow version.

```java
import java.io.File;
import java.lang.reflect.Field;
import java.lang.reflect.Method;
import java.net.MalformedURLException;
import java.net.URL;
import java.net.URLClassLoader;
import java.util.Vector;

public class ForceUnload2 {

    public static void main(String[] args) throws Exception {
        // Create ClassLoader with wanted JARs
        String file1 = "/home/carlos/Documents/test/tensorflow_0.3.3_linux_cpu/";
        URLClassLoader cl1 = new URLClassLoader(getURLs(file1));
        // Load native library of Tensorflow
        Class<?> ca1 = cl1.loadClass("org.bytedeco.javacpp.Loader");
        Class<?> ca3 = cl1.loadClass("org.tensorflow.internal.c_api.global.tensorflow");
        Method mm1 = ca1.getMethod("load", Class.class);
        mm1.invoke(null, ca3);
        // REmove references and try to garbage collect
        ca1 = null;
        mm1 = null;
        cl1.close();
        cl1 = null;
        for (int i = 0; i < 10; i++)
        	System.gc();

        System.gc();
		
        // Create ClassLoader for the second Tensorflow version
        String file = "/home/carlos/Documents/test/tensorflow_0.2.0_linux_cpu";
        URLClassLoader cl = new URLClassLoader(getURLs(file));
        // Load the second native library
        Class<?> ca = cl.loadClass("org.bytedeco.javacpp.Loader");
        Class<?> ca2 = cl.loadClass("org.tensorflow.internal.c_api.global.tensorflow");
        Method mm = ca.getMethod("load", Class.class);
        mm.invoke(null, ca2);
        // REmove all references to garbage collect
        ca2 = null;
        ca = null;
        mm = null;
        cl.close();
        cl = null;
        for (int i = 0; i < 10; i++)
        	System.gc();
	}
	
	public static URL[] getURLs(String file) throws MalformedURLException {
		File[] fs = new File(file).listFiles();
		URL[] urls = new URL[fs.length];
		int c = 0;
		for (File f : fs) {
			urls[c ++] = f.toURI().toURL();
		}
		return urls;
	}
}
```

I get the following error:
```
2022-01-06 13:40:14.640453: F external/org_tensorflow/tensorflow/core/framework/variant_op_registry.cc:46] Check failed: existing == nullptr (0x7f11b05451a8 vs. nullptr)Unary VariantDecodeFn for type_name: tensorflow::data::WrappedDatasetVariant already registered
```
It seems that the native library might have not been unloaded. Do you have any idea about how to solve this issue?
Regards,
Carlos

**Reply by carlosuc3m** [reply]
Oh sorry just saw your answers in the issue:
https://github.com/tensorflow/java/issues/402
My question is then, do you know if the same will happen for Pytorch?

**Reply by saudet** [reply]
Yes, I've confirmed that your method to unload works fine with PyTorch. However, we need to make sure that the deallocator thread from JavaCPP is not running when trying to do that, either by:
1. Setting the "org.bytedeco.javacpp.nopointergc" system property to "true" before loading JavaCPP
2. Calling `Pointer.interruptDeallocatorThread()` that I've just added for this purpose in commit https://github.com/bytedeco/javacpp/commit/3c9d5999b68367a20b9fd51a0c6965448aab6c61

**Reply by carlosuc3m** [reply]
Do I need to set "org.bytedeco.javacpp.nopointergc" system property to anyhing later?
Thanks for the fix! Will this work with any other native library?

**Reply by saudet** [reply]
I'm not sure what you mean, "org.bytedeco.javacpp.nopointergc" can only be set once:
http://bytedeco.org/javacpp/apidocs/org/bytedeco/javacpp/Pointer.html#referenceQueue
If you plan on calling `Pointer.interruptDeallocatorThread()` though, we don't need to set "org.bytedeco.javacpp.nopointergc". 
Yes, it works the same for other native libraries, as long as they support being unloaded.

**Reply by carlosuc3m** [reply]
Do I need to recompile the Pytorch Javacpp presets with the new JAvacpp version?

**Reply by saudet** [reply]
No, you can use the snapshots: http://bytedeco.org/builds/

**Reply by carlosuc3m** [reply]
Hello again @saudet .
Ive been trying the new presets and I am still finding some trouble. First of all when I load the pytorch presets with the following code:
```
try (URLClassLoader cls1 = new URLClassLoader(getURLs(file1))) {
    // Load native library of Pytorch
    Class<?> ca3 = cls1.loadClass("org.bytedeco.pytorch.Module");
    ca3.newInstance();
    Class<?> dealocateC = cls1.loadClass("org.bytedeco.javacpp.Pointer");
    Method dM = dealocateC.getMethod("interruptDeallocatorThread");
    dM.invoke(null);
    cls1.close();
    // REmove references and try to garbage collect
    ca3 = null;
}
```

I get the following warning:

```
OpenJDK 64-Bit Server VM warning: You have loaded library /home/carlos/.javacpp/cache/pytorch-1.10.1-1.5.7-20220104.184837-44-linux-x86_64.jar/org/bytedeco/pytorch/linux-x86_64/libtorch_cpu.so which might have disabled stack guard. The VM will try to fix the stack guard now.
It's highly recommended that you fix the library with 'execstack -c <libfile>', or link it with '-z noexecstack'.
```

But then, when I tried to load another version of the `libtorch_cpu.so` just with a simple `System.load()` I get the following error:
```java: ../nptl/pthread_mutex_lock.c:81: __pthread_mutex_lock: Assertion `mutex->__data.__owner == 0' failed.```

I have to say that I have been able to load and unload the Pytorch dlls in Windows relatively easy but never in Linux.
What do you think?
Regards and thanks a lot for your help,
Carlos

**Reply by saudet** [reply]
Indeed, it doesn't appear to be possible to reliably unload libraries on Linux:
https://stackoverflow.com/questions/24467404/dlclose-doesnt-really-unload-shared-object-no-matter-how-many-times-it-is-call

**Comment by saudet** [comment]
It should work more or less automatically with 1.8.x. You'll need to change the version in the `cppbuild.sh` and `pom.xml` files, then after removing `src/gen` try to run `mvn install` on that, and you'll probably get a couple of compiler errors, but it should be all fixable by adjusting `presets/torch.java` a bit. Let me know if you encounter any errors that you are not able to resolve though.

**Reply by carlosuc3m** [reply]
Hello again @saudet , first of all congrats on the awesome job you have done with this project. I have got to know it during the past few days and I am amazed.
I am very intereseted in loading all the machine learning frameworks (Tf, Pytorch, Mxnet, Onnx...) on my application using the Javacpp presets sooner or later (depending on the work it requires). As I need the different versions of each Deep LEarning framework available I am probably going to create presets for every version. Are the presets for Windows and Linux or GPU and CPU created at the same time? Or do I have to create them separately?
With respect to the dependencies specified in the Pytorch presets readme:
```
<project>
    <modelVersion>4.0.0</modelVersion>
    <groupId>org.bytedeco.pytorch</groupId>
    <artifactId>simplemnist</artifactId>
    <version>1.5.7-SNAPSHOT</version>
    <properties>
        <exec.mainClass>SimpleMNIST</exec.mainClass>
    </properties>
    <dependencies>
        <dependency>
            <groupId>org.bytedeco</groupId>
            <artifactId>pytorch-platform</artifactId>
            <version>1.10.1-1.5.7-SNAPSHOT</version>
        </dependency>

        <!-- Additional dependencies required to use CUDA, cuDNN, and NCCL -->
        <dependency>
            <groupId>org.bytedeco</groupId>
            <artifactId>pytorch-platform-gpu</artifactId>
            <version>1.10.1-1.5.7-SNAPSHOT</version>
        </dependency>

        <!-- Additional dependencies to use bundled CUDA, cuDNN, and NCCL -->
        <dependency>
            <groupId>org.bytedeco</groupId>
            <artifactId>cuda-platform-redist</artifactId>
            <version>11.4-8.2-1.5.7-SNAPSHOT</version>
        </dependency>

        <!-- Additional dependencies to use bundled full version of MKL -->
        <dependency>
            <groupId>org.bytedeco</groupId>
            <artifactId>mkl-platform-redist</artifactId>
            <version>2022.0-1.5.7-SNAPSHOT</version>
        </dependency>
    </dependencies>
    <build>
        <sourceDirectory>.</sourceDirectory>
    </build>
</project>
```
I guess that the `cuda-platorm-redist` are needed for GPU support, but what are the `mkl-platform-redist`?
Regards,
Carlos

**Note**
I am still asking this question in the original thread because I supposed it fits the title of the discussion

**Reply by saudet** [reply]
> Hello again @saudet , first of all congrats on the awesome job you have done with this project. I have got to know it during the past few days and I am amazed. I am very intereseted in loading all the machine learning frameworks (Tf, Pytorch, Mxnet, Onnx...) on my application using the Javacpp presets sooner or later (depending on the work it requires). As I need the different versions of each Deep LEarning framework available I am probably going to create presets for every version. Are the presets for Windows and Linux or GPU and CPU created at the same time? Or do I have to create them separately? With respect to the dependencies specified in the Pytorch presets readme:

Everything gets built with GitHub Actions more or less "at the same time", yes:
https://github.com/bytedeco/javacpp-presets/actions
> I guess that the `cuda-platorm-redist` are needed for GPU support, but what are the `mkl-platform-redist`? Regards, Carlos

The "redist" artifacts are there just for convenience, they are not required, no.

> **Note** I am still asking this question in the original thread because I supposed it fits the title of the discussion

I'm not sure what the question is anymore, but that's fine. I guess this discussion should be in the "General" category instead of "Q&A" though.

**Reply by carlosuc3m** [reply]
Great thank you for the info. The way I am doing it is forking your repo and changing the pom.xml files and ´cppbuild.sh´ to point to a different version of Pytorch. Then execute the Pytorch Github action. I guess that the file I might need to change is:
https://github.com/bytedeco/javacpp-presets/blob/master/pytorch/src/main/java/org/bytedeco/pytorch/presets/torch.java

Do you need I will need to change anything in the pytorch.yaml action, or in the workflows or in the cppbuild.sh, maybe to change the CUDA dependencies or something like that?
REgards,
CArlos

**Reply by saudet** [reply]
Possibly, but not necessarily, there are some differences, not everything is always different, that's why there are versions!

**Reply by carlosuc3m** [reply]
Great! Thank you for all the help during the past few days! I might come back to ask  a couple of questions :)
Regards,
CArlos

**Reply by saudet** [reply]
BTW, people usually do it the other way. They export models to a single new version of a runtime, which at the moment is centered around ONNX and ONNX Runtime, like those here from Hugging Face: https://huggingface.co/models

TVM is also an option, if you're willing to hack a bit more: http://bytedeco.org/news/2020/12/12/deploy-models-with-javacpp-and-tvm/

**Reply by carlosuc3m** [reply]
Yes I am familiar with that. However our application is oriented towards naive users so the version management should be done in the backend.

**Reply by carlosuc3m** [reply]
By the way Github actions cannot execute my build because it's longer than 6 hours, is there any way to alleviate it?
